src/volatile_memory.rs

```diff
@@ -1418,6 +1417,28 @@ mod copy_slice_impl {
 
         total
     }
+
+    /// Copies `total` bytes from `slice` to `dst`
+    pub(super) fn copy_from_volatile_slice<B: BitmapSlice>(
+        dst: *mut u8,
+        slice: &VolatileSlice<'_, B>,
+        total: usize,
+    ) -> usize {
+        // SAFETY: `slice` and `dst` must be point to a contiguously allocated memory region of at
+        // least length `total`. The regions must not overlap.
```

Comment:
Hmm, I normally add `unsafe` only to routines which directly contain `unsafe` rust code. This makes the users of the routine clean of the `unsafe` mess. I am not sure if this should be marked `unsafe` or not now.

@lauralt @andreeaflorescu any comments ?

Reply:
> Hmm, I normally add `unsafe` only to routines which directly contain `unsafe` rust code.

But this allows completely unsafe-free code (the user of the function) to trigger unsafe behavior unless your function prevents any case of unsafe behavior. In this example, the caller of this function is responsible for making sure that the conditions are verified. Declaring the function unsafe makes that explicit and easier to review.

`unsafe` is more than just a way to turn off checks in the compiler. It also is part of the function interface that helps the user write safe code. For example, many of the `unchecked` functions [1] do technically do anything that _requires_ an unsafe, but they are still declared unsafe, because they would otherwise easily allow introducing undefined/unsafe behaviour.

[1] https://doc.rust-lang.org/src/alloc/string.rs.html#845

> This makes the users of the routine clean of the `unsafe` mess.

I agree that things may be messy for functions that are used very commonly in the code. But if an unsafe function is so fundamental that it is used everywhere, then one should rather take it as a hint that this might be a candidate for creating a safe wrapper around it. I am not sure how practical that would be in this case, but one could, for example wrap the pointer with something that also holds an "allowed access range". Then a _safe_ function could check that the access stays within range and then perform the access using an _unsafe_ block without needing to turn the entire function unsafe.

Reply:
I was already updating the code with `unsafe` when your comments came. Done now.